Build compatible extensions

benjamin-confino opened the pull request:
Fixes #25977
Fixes #25978

What this does: 

- Moves the code that reads the CDI extension SPI out of CDIContainerImpl and into implementations of ExtensionArchiveProvider (EE7 through 9 redirect to common code in a Util class, EE10 does not)
- Modifies LiteExtensionProvider so it also processes SPIs. The end result is an extension archive that can contain both regular and build compatible extensions; as well as extra classes and annotations. Everything co-inhabits nicely. 
- Creates a new interface ExtensionContainer, if an ExtensionArchive contains one it will treat the extensions within like any extensions in an ExtensionArchive. This is a hack because the ExtensionArchive interface uses javax Extensions but LiteExtensionTranslator uses jakarta Extensions. 
- Splits the FAT test for the SPI into two. This prevents the original from using `cdi.internals-3.0` which would change what is being tested. `cdi.internals-3.0` is used on EE10 because the SPI for BCE currently is limited to our internal interface. (This was a surprisingly cumbersome process).

benjamin-confino committed:
Update SPI to support build compatible extensions

benjamin-confino commented on dev/io.openliberty.cdi.4.0.internal.weld/src/io/openliberty/cdi40/internal/weld/LiteExtensionArchive.java:
```diff
@@ -51,6 +52,10 @@ public LiteExtensionArchive(CDIRuntime cdiRuntime, ClassLoader bceClassLoader, L
         this.buildCompatibleExtensions = buildCompatibleExtensions;
     }
 
+    public LiteExtensionArchive(CDIRuntime cdiRuntime, ClassLoader bceClassLoader, Set<Class<? extends BuildCompatibleExtension>> buildCompatibleExtensions) {
```
Remove

benjamin-confino commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/archive/liberty/ExtensionArchiveImpl.java:
```diff
@@ -97,6 +102,29 @@ public Set<Supplier<Extension>> getSPIExtensionSuppliers() {
+                        maybeExtension = supplier.get();
+                        if (! (maybeExtension instanceof Extension)) {
+                            throw new IllegalArgumentException(maybeExtension.getClass().getCanonicalName()
+                                                           + " was registered as an extension via the WebSphereCDIExtensionMetaData interface. But it does not implement javax.enterprise.inject.spi.Extension");
```
Change this and previous to javax/jakarta

benjamin-confino committed:
Create tests for Build compatible extensions via SPI

benjamin-confino committed:
Use discovery mode none for runtime extensions with no beans.xml

benjamin-confino committed:
clean up and add test for SPI bundle with no extensions

benjamin-confino commented:
#build

Azquelt commented:
The last two commits appear to be a separate change which needs its own issue and its own PR, particularly as it should be much simpler and easier to review and merge than this one.

Azquelt commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/interfaces/ExtensionArchiveProvider.java:
```diff
@@ -29,6 +38,11 @@ public interface ExtensionArchiveProvider {
      * @return a collection of additional extension archives to add
      * @throws CDIException if an unexpected exception is encountered
      */
-    Collection<ExtensionArchive> getArchives(CDIRuntime cdiRuntime, WebSphereCDIDeployment deployment) throws CDIException;
+    default Collection<ExtensionArchive> getDiscoveredArchives(CDIRuntime cdiRuntime, WebSphereCDIDeployment deployment) throws CDIException {
+        return Collections.emptySet();
+    }
```
This name change doesn't make sense. The archives added by implementors of this interface are not "discovered" - they're created anew.

benjamin-confino replied:
I was thinking of the fact we LiteExtensionProvider was creating archives for BCEs discovered through the classloader. I have no objection to changing the name though.

Azquelt commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/interfaces/ExtensionArchiveProvider.java:
```diff
@@ -29,6 +38,11 @@ public interface ExtensionArchiveProvider {
+        return Collections.emptySet();
+    }
+
+    ExtensionArchive newSPIExtensionArchive(CDIRuntime cdiRuntime, ServiceReference<CDIExtensionMetadata> sr,
+                                                    CDIExtensionMetadata webSphereCDIExtensionMetaData, WebSphereCDIDeployment applicationContext) throws CDIException;
```
This method doesn't make sense on this class. Please put it on its own interface.

benjamin-confino replied:
Do we really need two interfaces for version specific code for generating extension archives? Perhaps rather than splitting the interface we should find a framing where it makes sense to have both methods?

Azquelt replied:
As discussed, I think we do, because we need exactly one implementation of this method, whereas we could have any number of the other method.

Azquelt commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/interfaces/ExtensionArchive.java:
```diff
@@ -43,4 +43,7 @@ public interface ExtensionArchive extends CDIArchive {
      */
     Set<Supplier<Extension>> getSPIExtensionSuppliers();
+
+    default void addLiteExtensionTranslator(Extension liteExtensionTranslator, Set<String> buildCompatibleExtensionClassNames) {};
```
This method is weird and takes a completely different approach to everything else on this interface. Every other option is set via the constructor, why would this one have a setter?

Also, like every method on an interface like this, it should have Javadoc.

Azquelt replied:
Initially, I thought there was only one implementation of this interface so I was confused as to why we have two: the general purpose `ExtensionArchiveImpl` and the special case `ProbeExtensionArchive`.

This method would only work on `ExtensionArchiveImpl`, so I'm becoming more and more convinced it shouldn't be on this interface.

Azquelt replied:
Furthermore, it doesn't strictly need to be a Lite Extension Translator extension, right? You're just adding a list of class names to be reported as "extension classes" and an extension instance you'd like to be added to the extension archive.

benjamin-confino replied:
You're correct that it doesn't strictly need to be a Lite Extension Translator. 

We could call it something like addConstructedExtension instead. As this method exists because we need to pass in Lite Extension Translator objects rather than just a class name.

benjamin-confino replied:
It uses a setter simply because I thought it would be less ugly to call a setter from cdi 4.0 ExtensionArchiveProvider than to have the other three pass in nulls. 

If you'd prefer to use the constructor that would be an easy change.

Azquelt replied:
I think `addConstructedExtension` is a good name.

As discussed, I think this does need to come off the interface because we have three implementations (I forgot `LiteExtensionArchive`) and only one will ever implement it.

I don't mind whether we move this to the constructor (which is a bit unwieldy because it has two arguments) or cast to the implementation to call the method.

Azquelt commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/archive/liberty/ExtensionArchiveImpl.java:
```diff
@@ -70,6 +78,7 @@ public boolean isExtClassesOnly() {
     public Set<String> getExtensionClasses() {
         Set<String> extensionClasses = super.getExtensionClasses();
         extensionClasses.addAll(spiExtensions);
+        extensionClasses.addAll(spiBuildCompatibleExtensions);
         return extensionClasses;
     }
```
What is `getExtensionClasses()` used for?

Is it correct to have build compatible extension classes in here?

Azquelt replied:
Confusingly, there is also a method `CDIArchive.getBuildCompatibleExtensionClasses()`. However, it's currently used to determine that we need to create a lite extension translator for an application bundle.

benjamin-confino replied:
`getExtensionClasses` is used by BDAFactory to pass in the extensionClassNames to BeanDeploymentArchiveImpl's constructor. After that there's code in scanForBeanClassNames that can include it in `beanClasses`. Which gets passed up to weld via `getBeanClasses`

I've read both the javadoc and the integration guide for what weld expects `getBeanClasses` and it was pretty terrible, but it looks like it expects results based on the bean scanning mode which is defined in the CDI spec, and checking the 4.0 spec I didn't see any reason to treat BCE and regular Es differently.

Azquelt replied:
As discussed, we don't think the spiBuildCompatibleExtensions should be added here.

However, we might need the `LiteExtensionTranslator` class added. I think some parts of BeanDeploymentArchiveImpl assume we can tell if an archive is an extension archive by seeing if it has either an Extension class or a BCE class.

Azquelt commented on dev/io.openliberty.cdi.4.0.internal.weld/src/io/openliberty/cdi40/internal/archive/liberty/package-info.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * @version 1.0
+ */
+@org.osgi.annotation.versioning.Version("1.0")
+package io.openliberty.cdi40.internal.archive.liberty;
```
Why is this added?

benjamin-confino replied:
Remember the previous version with container classes. A left over from that which can be safely removed.

Azquelt commented on dev/io.openliberty.jakarta.cdi.3.0/src/com/ibm/ws/cdi/impl/CDI30ExtensionSPIProvider.java:
```diff
@@ -0,0 +1,38 @@
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.framework.ServiceReference;
+
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class CDI30ExtensionSPIProvider implements ExtensionArchiveProvider {
```
This shouldn't go in the API bundle

Azquelt replied:
Turns out we don't need this class at all because the `com.ibm.ws.cdi.2.0.weld` bundle is also included in the CDI 3.0 feature.

Azquelt commented on dev/com.ibm.ws.cdi.interfaces/src/com/ibm/ws/cdi/extension/CDIExtensionMetadataInternal.java:
```diff
@@ -12,6 +12,9 @@
  *******************************************************************************/
 package com.ibm.ws.cdi.extension;
 
+import java.util.Collections;
+import java.util.Set;
+
 import io.openliberty.cdi.spi.CDIExtensionMetadata;
 
```
No real changes in this file

benjamin-confino commented:
> The last two commits appear to be a separate change which needs its own issue and its own PR, particularly as it should be much simpler and easier to review and merge than this one.

They're in the same commit because I want to make sure the big changes in the earlier commits pass the new tests. And I wanted to write the new tests in the same branch to avoid a merge conflict with the tests from the earlier commit.

benjamin-confino committed:
code review changes

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/test-bundles/cdi.spi.extension/src/com/ibm/ws/cdi/extension/spi/test/bundle/buildcompatible/BCExtraBean.java:
```diff
@@ -0,0 +1,25 @@
+ *******************************************************************************/
+package com.ibm.ws.cdi.extension.spi.test.bundle.buildcompatible;
+
+//Even though this is only run in EE10 we use javax otherwise the MANIFEST.MF imports both javax and jakarta.enterprise.context; then blows up when the transformer doesn't filter out the duplicate
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class BCExtraBean {
```
I don't think this class is used anywhere

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi/extension/apps/spi/SPIExtensionServlet.java:
```diff
@@ -23,6 +23,7 @@
 import org.junit.Test;
 
 import com.ibm.ws.cdi.extension.spi.test.bundle.UnregisteredBean;
+import com.ibm.ws.cdi.extension.spi.test.bundle.buildcompatible.BCExtraBean;
```
No real changes in this file.

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi/extension/apps/spi40/SPIExtensionServlet.java:
```diff
@@ -0,0 +1,109 @@
+@WebServlet("/spi")
+public class SPIExtensionServlet extends FATServlet {
+
+    @Inject
+    MyExtensionString extensionString;
+
+    @Inject
+    MyProducedString classString;
+
+    @Inject
+    MyBeanInjectionString beanInjectedString;
+
+    @Inject
+    AppBean appBean;
+
+    @Inject
+    BCExtraBean bcExtraBean;
+
+    @Inject
+    CustomBDABean customBDABean;
```
Should really make all of these private

Azquelt commented on dev/io.openliberty.cdi.4.0.interfaces/bnd.bnd:
```diff
@@ -0,0 +1,30 @@
+#*******************************************************************************
+-include= ~../cnf/resources/bnd/bundle.props
+
+bVersion=1.0
+	com.ibm.ws.cdi.extension
```
The bnd file is missing the bundle name and symbolic name and the project is missing the eclipse project files.

I'll make a commit to fix both of these.

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi/extension/apps/spi40/SPIExtensionServlet.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    @Test
+    public void testSPIInterceptorNormal() {
+        assertEquals("Produced injection", classString.toString());
+    }
```
This test looks like it tests a producer, not an interceptor.

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi/extension/tests/CDI12NoExtensionSPITest.java:
```diff
@@ -0,0 +1,81 @@
+        if (server.isStarted()) {
+            server.stopServer();
+        }
+        Log.info(CDI12ExtensionTest.class, METHOD_NAME, "Removing spi extension test user feature files.");
+        CDIExtensionRepeatActions.uninstallUserExtension(server, CDIExtensionRepeatActions.CDI_SPI_WITH_NO_EXTENSION_BUNDLE_ID);
```
Should go in a try-finally block to ensure we always remove the user feature

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi/extension/tests/CDI40ExtensionSPITest.java:
```diff
@@ -0,0 +1,93 @@
+ * Rather than enable access to internals for everything I split the test so we can test with and without that access.
+ */
+@RunWith(FATRunner.class)
+@Mode(TestMode.LITE)
+public class CDI40ExtensionSPITest extends FATServletClient {
```
As discussed, this can be a FULL mode test.

Azquelt committed:
Update Eclipse project files

Add project files for io.openliberty.cdi.4.0.interfaces and fix up bnd
file.

Add source folder in com.ibm.ws.cdi.extension_fat

Azquelt commented:
The other thing missing was copyright dates on some modified files.

benjamin-confino committed:
code review changes

benjamin-confino committed:
copyright dates

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi/extension/tests/CDIExtensionRepeatActions.java:
```diff
@@ -82,7 +84,7 @@ public static String getFeatureName(String bundleID, EEVersion eeVersion) {
         } else if (eeVersion == EEVersion.EE9) {
             name = bundleID + "-3.0";
         } else if (eeVersion == EEVersion.EE10) {
-            name = bundleID + "-3.0";
+            name = bundleID + "-4.0";
```
Update copyright date on this file.

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/fat/src/com/ibm/ws/cdi/extension/tests/FATSuite.java:
```diff
@@ -28,10 +28,11 @@
                 CDI12ExtensionSPITest.class,
+                CDI12NoExtensionSPITest.class,
+                CDI40ExtensionSPITest.class,
                 DynamicBeanExtensionTest.class,
                 ObserverTest.class,
-                WeldDevModeTest.class
-})
+                WeldDevModeTest.class})
```
Update copyright date on this file

Azquelt commented on dev/com.ibm.ws.cdi.extension_fat/test-bundles/cdi.spi.extension/src/com/ibm/ws/cdi/extension/spi/test/bundle/SPIMetaData.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2020. 2021 IBM Corporation and others.
```
Undo date change, file has not changed.

Azquelt commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/interfaces/ExtensionArchiveProvider.java:
```diff
@@ -29,6 +33,7 @@ public interface ExtensionArchiveProvider {
      * @return a collection of additional extension archives to add
      * @throws CDIException if an unexpected exception is encountered
      */
-    Collection<ExtensionArchive> getArchives(CDIRuntime cdiRuntime, WebSphereCDIDeployment deployment) throws CDIException;
-
+    default Collection<ExtensionArchive> getArchives(CDIRuntime cdiRuntime, WebSphereCDIDeployment deployment) throws CDIException {
+        return Collections.emptySet();
+    }
```
This change is no longer needed since it's the only method on the interface.

The rest of the changes to this file (copyright date, imports) should also be reverted.

Azquelt commented on dev/io.openliberty.cdi.4.0.internal.weld/src/io/openliberty/cdi40/internal/weld/CDI40ExtensionSPIProvider.java:
```diff
@@ -12,35 +12,54 @@
 @Component(configurationPolicy = ConfigurationPolicy.IGNORE)
-public class LiteExtensionProvider implements ExtensionArchiveProvider {
-    private static final TraceComponent tc = Tr.register(LiteExtensionProvider.class);
+public class CDI40ExtensionSPIProvider implements ExtensionArchiveProvider, ExtensionArchiveFactory {
```
Can you move the class javadoc to the `getArchives()` method please? It only really applies to that method.

Azquelt commented on dev/com.ibm.ws.cdi.interfaces/src/io/openliberty/cdi/spi/CDIExtensionMetadata.java:
```diff
@@ -68,6 +68,8 @@ public interface CDIExtensionMetadata {
      * All classes returned by this method will be will be found by CDI during type discovery
      * so that they can then be used as beans (or interceptors etc. if annotated as such) by the application.
      * All classes must be in the same archive as your CDIExtensionMetadata.
+     *
+     * If a bean registered via this does not have a bean defining annotation it will default to a scope of @Dependent
      */
```
As this is SPI, the Javadoc is published so we should format it properly.
```suggestion
     * All classes must be in the same archive as your CDIExtensionMetadata.
     * <p>
     * If a bean registered via this does not have a bean defining annotation it will default to a scope of {@code @Dependent}
     */
```

Azquelt commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/archive/liberty/ExtensionArchiveImpl.java:
```diff
@@ -29,11 +29,19 @@
 public class ExtensionArchiveImpl extends CDIArchiveImpl implements ExtensionArchive {
 
-    private static final TraceComponent tc = Tr.register(ExtensionArchiveImpl.class);
+    protected static final TraceComponent tc = Tr.register(ExtensionArchiveImpl.class);
```
Why does this need to be `protected`? Usually a subclass would have its own `TraceComponent`.

Azquelt commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/archive/liberty/ExtensionArchiveImpl.java:
```diff
@@ -34,6 +34,14 @@ public class ExtensionArchiveImpl extends CDIArchiveImpl implements ExtensionArc
     private final ExtensionContainerInfo extensionContainerInfo;
     private Set<String> spiExtensions = null;
 
+    //Store these seperately so we don't try to make regular extensions out of them.
+    private Set<String> preConstructedExtensionNames = new HashSet<String>();
```
We add class names to this, but we never retrieve them.

benjamin-confino commented:
#build

benjamin-confino commented:
#build

benjamin-confino committed:
export new interface bundle

benjamin-confino commented:
#build

tevans78 reviewed (approved):
@Azquelt already approved the first 8 commits. I'm approving the last one.

benjamin-confino commented:
previous build was at commit b77e43bbb379399040d2aca2a87e8d423ac73d8c, before the feature update to export the new interface

benjamin-confino commented:
#build

benjamin-confino commented:
Only failures were in com.ibm.ws.jpa.tests.container_modifyconfig_fat FAT tests. I see injection but it looks like this bucket has no CDI injection.

benjamin-confino commented:
#libby
